pkg/ibmcsidriver/node.go

```diff
@@ -543,10 +550,16 @@ func (su *VolumeStatUtils) IsDevicePathNotExist(devicePath string) bool {
 }
 
 // Resize expands the fs
-func (volMountUtils *VolumeMountUtils) Resize(mounter mountmanager.Mounter, devicePath string, deviceMountPath string) (bool, error) {
-	r := mount.NewResizeFs(mounter.NewSafeFormatAndMount().Exec)
-	if _, err := r.Resize(devicePath, deviceMountPath); err != nil {
-		return false, err
+func (volMountUtils *VolumeMountUtils) Resize(mounter *mount.SafeFormatAndMount, devicePath string, deviceMountPath string) (bool, error) {
+	r := mount.NewResizeFs(mounter.Exec)
+	needResize, err := r.NeedResize(devicePath, deviceMountPath)
+	if err != nil {
```

Comment:
Resizing in `NodeExpandVolume()` is not the issue here. We need to do the resize in `NodeStageVolume()`. This is because when you restore from a snapshot to a new PVC the expand function never gets called since this is not expand operation in fact. In other words we need to check the filesystem during staging and resize if needed.

If you read my commit messages from first to last it should explain different parts of the changes - the change here is a separate commit that just improves resizing to be "smarter" and do the resize only if needed. It has nothing to do with the original bug.